src/pkg/scan/dao/scan/vulnerability.go

```diff
@@ -0,0 +1,230 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scan
+
+import (
+	"context"
+	"fmt"
+	"github.com/goharbor/harbor/src/lib/orm"
+	"github.com/goharbor/harbor/src/lib/q"
+)
+
+func init() {
+	orm.RegisterModel(new(VulnerabilityRecord), new(ReportVulnerabilityRecord))
+}
+
+type VulnerabilityRecordDao interface {
+	// Create creates a new vulnerability record
+	Create(ctx context.Context, vr *VulnerabilityRecord) (int64, error)
+	// Delete deletes a vulnerability record
+	Delete(ctx context.Context, vr *VulnerabilityRecord) error
+	// List lists the vulnerability records
+	List(ctx context.Context, query *q.Query) ([]*VulnerabilityRecord, error)
+	// InsertForReport inserts vulnerability records for a report
+	InsertForReport(ctx context.Context, reportUUID string, vr *VulnerabilityRecord) (int64, error)
+	// GetForReport gets vulnerability records for a report
+	GetForReport(ctx context.Context, reportUUID string) ([]*VulnerabilityRecord, error)
+	// GetForScanner gets vulnerability records for a scanner
+	GetForScanner(ctx context.Context, registrationUUID string) ([]*VulnerabilityRecord, error)
+	// DeleteForScanner deletes vulnerability records for a scanner
+	DeleteForScanner(ctx context.Context, registrationUUID string) (int64, error)
+	// DeleteForReport deletes vulnerability records for a report
+	DeleteForReport(ctx context.Context, reportUUID string) (int64, error)
+	// DeleteForDigests deletes vulnerability records for a provided list of digests
+	DeleteForDigests(ctx context.Context, digests ...string) (int64, error)
+	// GetRecordIdsForScanner gets record ids of vulnerability records for a scanner
+	GetRecordIdsForScanner(ctx context.Context, registrationUUID string) ([]int, error)
+}
+
+// NewVulnerabilityRecordDao returns a new dao to handle vulnerability data
+func NewVulnerabilityRecordDao() VulnerabilityRecordDao {
+	return &vulnerabilityRecordDao{}
+}
+
+type vulnerabilityRecordDao struct{}
+
+// Create creates new vulnerability record.
+func (v *vulnerabilityRecordDao) Create(ctx context.Context, vr *VulnerabilityRecord) (int64, error) {
+	o, err := orm.FromContext(ctx)
+	if err != nil {
+		return 0, err
+	}
+	_, vrID, err := o.ReadOrCreate(vr, "CVEID", "RegistrationUUID", "Digest", "Package", "PackageVersion")
```

Comment:
For some reason the "Digest" property lead to the problem that not all vulnerability records were persisted in the database. After removing if from the model I'm able to calculate summary and list vulnerability items for the specified artifact digest and mime type with the following queries:

```sql
-- Find severity summary for the specified artifact digest and report mime type
SELECT vr.severity, count(vr.severity) FROM vulnerability_record vr
  INNER JOIN report_vulnerability_record rvr ON vr.id = rvr.vuln_record_id
  INNER JOIN scan_report sr ON sr.uuid = rvr.report_uuid
  WHERE sr.digest = 'sha256:07fffd786a81e5f23fe4f5f58a181b43084ed232e1db732415023848937b9c73'
  AND sr.mime_type = 'application/vnd.scanner.adapter.vuln.report.harbor+json; version=1.0'
  -- and sr.mime_type = 'application/vnd.security.vulnerability.report; version=1.1'
  GROUP BY vr.severity;
```

```sql
-- Find vulnerabilities for the specified artifact digest and report mime type
SELECT vr.* FROM vulnerability_record vr
  INNER JOIN report_vulnerability_record rvr ON vr.id = rvr.vuln_record_id
  INNER JOIN scan_report sr ON sr.uuid = rvr.report_uuid
  WHERE sr.digest = 'sha256:07fffd786a81e5f23fe4f5f58a181b43084ed232e1db732415023848937b9c73'
  AND sr.mime_type = 'application/vnd.scanner.adapter.vuln.report.harbor+json; version=1.0'
  -- and sr.mime_type = 'application/vnd.security.vulnerability.report; version=1.1'
  ;
```
```suggestion
	_, vrID, err := o.ReadOrCreate(vr, "CVEID", "RegistrationUUID", "Package", "PackageVersion")
```

Reply:
Done. I have removed the digest column and things should look good now.